numpy/lib/index_tricks.pyi

```diff
@@ -0,0 +1,179 @@
+import sys
+from typing import (
+    Any,
+    Tuple,
+    TypeVar,
+    Generic,
+    overload,
+    List,
+    Union,
+    Sequence,
+)
+
+from numpy import (
+    # Circumvent a naming conflict with `AxisConcatenator.matrix`
+    matrix as _Matrix,
+    ndenumerate as ndenumerate,
+    ndindex as ndindex,
+    ndarray,
+    dtype,
+    str_,
+    bytes_,
+    bool_,
+    int_,
+    float_,
+    complex_,
+    intp,
+    _OrderCF,
+    _ModeKind,
+)
+from numpy.typing import (
+    # Arrays
+    ArrayLike,
+    _NestedSequence,
+    _RecursiveSequence,
+    _ArrayND,
+    _ArrayOrScalar,
+    _ArrayLikeInt,
+
+    # DTypes
+    DTypeLike,
+    _SupportsDType,
+
+    # Shapes
+    _ShapeLike,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal, SupportsIndex
+else:
+    from typing_extensions import Literal, SupportsIndex
+
+_T = TypeVar("_T")
+_DType = TypeVar("_DType", bound=dtype[Any])
+_BoolType = TypeVar("_BoolType", Literal[True], Literal[False])
+_SliceOrTuple = TypeVar("_SliceOrTuple", bound=Union[slice, Tuple[slice, ...]])
+_ArrayType = TypeVar("_ArrayType", bound=ndarray[Any, Any])
+
+__all__: List[str]
+
+def unravel_index(
+    indices: _ArrayLikeInt,
+    shape: _ShapeLike,
+    order: _OrderCF = ...
+) -> Tuple[_ArrayOrScalar[intp], ...]: ...
+
+def ravel_multi_index(
+    multi_index: ArrayLike,
+    dims: _ShapeLike,
+    mode: Union[_ModeKind, Tuple[_ModeKind, ...]] = ...,
+    order: _OrderCF = ...
+) -> _ArrayOrScalar[intp]: ...
+
+@overload
+def ix_(*args: _NestedSequence[_SupportsDType[_DType]]) -> Tuple[ndarray[Any, _DType], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[str]) -> Tuple[_ArrayND[str_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[bytes]) -> Tuple[_ArrayND[bytes_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[bool]) -> Tuple[_ArrayND[bool_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[int]) -> Tuple[_ArrayND[int_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[float]) -> Tuple[_ArrayND[float_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[complex]) -> Tuple[_ArrayND[complex_], ...]: ...
+@overload
+def ix_(*args: _RecursiveSequence) -> Tuple[_ArrayND[Any], ...]: ...
```

Comment:
These annotations seem odd to me - `ix_` just has an `ArrayLike` in each `args[i]`. Is there no way at the moment to type this as something like
```
def ix_(*args: ArrayLike[T]) -> Tuple[_ArrayND[T], ...]: ...
```

Reply:
The big challenge here is that we'd somehow have to express the builtin scalar types as some object parametrized w.r.t. `np.dtype`/`np.generic`, via a protocol for example: `SuperFancyProtocol[np.int_] == builtins.int`.

The problem is that no such protocol or object exists, so we're stuck with the current situation.

This is even more of an issue with dtypes, as there are roughly ~30 sets of string-literals that each map to their own dtype.

Reply:
Fair enough - I haven't been following the typing too closely.

Reply:
This was the same question I had, so let me unresolve this - it has value for people looking at this in the future.

Also, this is a good reason to exclude `dtype=builtins.int`, format strings and the like in `numpy.array_api` (NEP 47). For all the functions in that namespace the `func([ndarray[T]]) -> ndarray[T]: ...` pattern should work.